npe2/_types.py

```diff
@@ -31,9 +40,28 @@ def __array__(self) -> "np.ndarray":
 DataType = Union[ArrayLike, Sequence[ArrayLike]]
 FullLayerData = Tuple[DataType, Metadata, LayerName]
 LayerData = Union[Tuple[DataType], Tuple[DataType, Metadata], FullLayerData]
-PathLike = Union[str, Path]
-PathOrPaths = Union[PathLike, Sequence[PathLike]]
+
+# ########################## CONTRIBUTIONS #################################
+
+# WidgetContribution.command must point to a WidgetCreator
+Widget = Union["magicgui.widgets.Widget", "qtpy.QtWidgets.QWidget"]
+WidgetCreator = Callable[..., Widget]
+
+# ReaderContribution.command must point to a ReaderGetter
 ReaderFunction = Callable[[PathOrPaths], List[LayerData]]
-WriterFunction = Callable[[str, List[FullLayerData]], List[str]]
-Widget = Union["FunctionGui", "QWidget"]
-WidgetCallable = Callable[..., Widget]
+ReaderGetter = Callable[[Union[str, List[str]]], Optional[ReaderFunction]]
+
+# SampleDataGenerator.command must point to a SampleDataCreator
+SampleDataCreator = Callable[..., List[LayerData]]
+
+# WriterContribution.command must point to a WriterFunction
+# Writers that take at most one layer must provide a SingleWriterFunction command.
+# Otherwise, they must provide a MultiWriterFunction.
+# where the number of layers they take is defined as
+# n = sum(ltc.max() for ltc in WriterContribution.layer_type_constraints())
```

Comment:
```suggestion
# WriterContribution.command must point to a WriterFunction.
# Currently, two calling conventions are supported for writers: single-layer and
# multi-layer writers. When at most one layer can be matched by a writer, it
# must use the single-layer convention. Otherwise, the multi-layer convention
# must be used. WriterFunctions are only called when the set of layers to be
# written satisfies a set of layer-type constraints specified by
# WriterContribution.layer_types.layer_type_constraints().
```

Reply:
try 2

Reply:
thx, merged.  i think what I was really looking for though, is something more explicit than "When at most one layer can be matched by a writer, it must use the single-layer convention. Otherwise, the multi-layer convention must be used.".  This docstring _still_ requires me to go elsewhere to figure it out.  it references `WriterContribution.layer_types.layer_type_constraints`, which also doesn't have a docstring... so I have to dig deeper. 

`n = sum(ltc.max() for ltc in WriterContribution.layer_type_constraints())` was a bit better, since it tells me _exactly_ what it means to match more than one layer ... but there too, I have to go through `LayerTypeConstraint.max` and `LayerTypeConstraint.bounds` before I finally get to something that tells me what the _manifest_ needs to do to indicate these things.  Let's do this in a follow up though